src/text/parsers/text_support.rs

```diff
@@ -62,28 +62,43 @@ pub(crate) fn escaped_char_literal(input: &str) -> IResult<&str, char> {
 /// Matches a Unicode escape (starting with '\x', '\u', or '\U'), returning the appropriate
 /// substitute character.
 pub(crate) fn escaped_char_unicode(input: &str) -> IResult<&str, char> {
-    map_res::<_, _, _, _, IonError, _, _>(
-        alt((
-            escaped_char_unicode_2_digit_hex,
-            escaped_char_unicode_4_digit_hex,
-            escaped_char_unicode_8_digit_hex,
-        )),
-        |hex_digits| {
-            let number_value = u32::from_str_radix(hex_digits, 16).map_err(|e| {
-                decoding_error_raw(format!(
-                    "Couldn't parse unicode escape '{}': {:?}",
-                    hex_digits, e
-                ))
-            })?;
-            let char_value = std::char::from_u32(number_value).ok_or_else(|| {
-                decoding_error_raw(format!(
-                    "Couldn't parse unicode escape '{}': {} is not a valid codepoint.",
-                    hex_digits, number_value
-                ))
-            })?;
-            Ok(char_value)
-        },
-    )(input)
+    // First, try to match the input to Unicode escape sequence. If successful, extract the hex
+    // digits that were included in the sequence. If matching fails, this isn't an escape sequence.
+    // Return early with a non-fatal error.
+    let (remaining_input, hex_digits) = alt((
+        escaped_char_unicode_2_digit_hex,
+        escaped_char_unicode_4_digit_hex,
+        escaped_char_unicode_8_digit_hex,
+    ))(input)?;
+
+    // Now that we have our hex digits, we'll try to convert them to Unicode code points.
+    // If this step fails, the Ion data stream is malformed and we need to bail out completely.
+    // We can't simply return an error as we did above; if we did that, the parser would go on to
+    // treat the input as a String literal without escapes, which is the incorrect behavior.
+    // Instead, we need to return a nom `Err::Failure`, indicating that we cannot proceed.
+    let number_value = match u32::from_str_radix(hex_digits, 16) {
+        Ok(number_value) => number_value,
+        Err(_parse_int_error) => {
+            // TODO: A custom error type would be required to bubble up specific information
+            //       about the failure.
+            return Err(Err::Failure(Error::from_error_kind(
```

Comment:
🗺️ Typically, when a `nom` parser returns an `Err` it's a signal that it didn't find what it was looking for in the input. This is both common and expected; the reader will move on to another parser and see if that one matches instead. However, there are times when the reader is definitely using the right parser but it has found a fatal problem. In this case, we've found a Unicode escape sequence that's correctly formatted (e.g. `\xE0`) but when we tried to turn it into a codepoint, we realized it wasn't valid. We need to communicate to the reader that the stream is malformed and it shouldn't continue on trying other parsers. In cases like this, we want to return `nom`'s `Err::Failure(_)` instead of `Err::Error(_)`, which is what this change does.

Unfortunately, the default `nom` error/failure type doesn't provide a way to bubble up contextual information about the problem it encountered to the reader. As a future task, we'll need to make a custom parsing error type that allows us to provide (e.g.) a `String` explaining what went wrong.